training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala

```diff
@@ -175,6 +176,73 @@ object FullRankLinearTrainer {
     .collect
   }
 
+  def hingeGradient(sc : SparkContext,
+                    options : FullRankLinearTrainerOptions,
+                    model : FullRankLinearModel,
+                    pointwise : RDD[Example]) : Array[((String, String), GradientContainer)] = {
+    val modelBC = sc.broadcast(model)
+
+    pointwise
+      .mapPartitions(partition => {
+        val model = modelBC.value
+        val labelToIdx = model.getLabelToIndex()
+        val dim = model.getLabelDictionary.size()
+        val gradient = scala.collection.mutable.HashMap[(String, String), GradientContainer]()
+        val weightVector = model.getWeightVector()
+        val rnd = new Random()
+
+        partition.foreach(examples => {
```

Comment:
partition consists of multiply examples, but in partition.foreach we are only looking at one example, so perhaps rename this as `example`